ping/ping.c

```diff
@@ -803,9 +836,13 @@ int ping4_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 			error(2, errno, "IP_MTU_DISCOVER");
 	}
 
-	if (rts->opt_strictsource &&
-	    bind(sock->fd, (struct sockaddr *)&rts->source, sizeof rts->source) == -1)
-		error(2, errno, "bind");
+	int do_need_to_set_ident = rts->ident > 0 && sock->socktype == SOCK_DGRAM;
+	if (rts->opt_strictsource || do_need_to_set_ident) {
+		if (do_need_to_set_ident)
+			rts->source.sin_port = rts->ident;
+		if (bind(sock->fd, (struct sockaddr *)&rts->source, sizeof rts->source) == -1)
+			error(2, errno, "bind");
+  }
```

Comment:
This part was not in original propose. I don't like the code style when variables are set before using, but it just follows the bad style from the past (including long `if/else` parts.
* `do_need_to_set_ident` is too long, how about just `set_ident`?
* final `}` has wrong indent.
IMHO this is a bit better:
```
int set_ident = rts->ident > 0 && sock->socktype == SOCK_DGRAM;
if (set_ident)
	rts->source.sin_port = rts->ident;

if (rts->opt_strictsource || set_ident) {
	if (bind(sock->fd, (struct sockaddr *)&rts->source, sizeof rts->source) == -1)
		error(2, errno, "bind");
}
```

Reply:
done